### PR TITLE
Menu fixes for play tones

### PR DESCRIPTION
**Pull request by cspiel1**
- ringback tone was not played
- fixes when and what to play for incoming calls
- resume right tone if a call terminates
- fix for alert_stop (terminal beep)

Has to be rebased after merge of https://github.com/baresip/baresip/pull/1272.

### CODE REVIEW

File: modules/menu/menu.c [hunk 1]
```diff
@@ -134,6 +134,14 @@ static char *errorcode_key_aufile(uint16_t scode)
 }
 
 
+static void menu_stop_play(void)
+{
+	menu.play = mem_deref(menu.play);
+	menu.ringback = false;
+	alert_stop();
```
Comment (alfredh) [inline modules/menu/menu.c:141]: unrelated comment; should we consider to remove the "alert" code ?

Comment (cspiel1) [reply modules/menu/menu.c:141]: Yes, I think so. Another PR or may add here a commit?

**Comment by cspiel1** [comment]
The alert beep:
- It is somehow related to this PR and already touched here.
- E.g. Does not work on default Ubuntu systems because beep seems to be deactivated.
- Alert is obsolete, since a WAV file can be played everywhere with baresip. So ringtones should work anyway.

Will add a commit to remove the alert code. Let me know if you have counter-arguments!

**Comment by cspiel1** [comment]
Appending a call to a custom list is not possible because calls are already tracked in the list of its UA. I tried a solution with this new API functions:
```
void call_append(struct list *calls, struct call *call);
int ua_filter_calls(struct ua *ua, struct list *calls, enum call_state st);
int uag_filter_calls(struct list *calls, enum call_state st);
```
But call_append() is not possible. It results in this error:
```
list: append: le linked to 0x55b50f9b5c80
```

I try to find a different solution for this ringtone PR. But you are right, that a call filter would be helpful in general.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Appending a call to a custom list is not possible because calls are
> already tracked in the list of its UA.

So a call object can only be in one list (ua's struct list calls)?  If
so, can it be relaxed somehow or does re make it impossible?

**Comment by cspiel1** [comment]
Yes, exactly. No it can't be relaxed. A list-element (struct le) has one pointer prev and one next.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Yes, exactly. No it can't be relaxed. A list-element (struct le) has
> one pointer prev and one next.

Then how about making a copy of the list elements?

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Yes, exactly. No it can't be relaxed. A list-element (struct le) has
> one pointer prev and one next.

List element has `void *data`.  How about creating a new list with new
elements and take the data to element from the ua calls list?

**Comment by cspiel1** [comment]
> Then how about making a copy of the list elements?
The call->le is hidden. A copy of a call, is too much.

But look, here I have two implemenations of menu_find_active_call(). The first one uses the call filter (which currently does not work). The second makes everything of its own.

```
static struct call *menu_find_active_call(void)
{
	struct list calls;
	struct le *le;
	int err;
	struct call *call = NULL;

	list_init(&calls);
	err = uag_filter_calls(&calls, CALL_STATE_ESTABLISHED);
	if (err)
		return NULL;

	for (le = calls.head; le; le = le->next) {
		call = le->data;

		if (!call_is_onhold(call))
			goto out;
	}

out:
	list_clear(&calls);
	return call;
}
```
Here without the call filter uag_filter_calls():
```
static struct call *menu_find_active_call(void)
{
	struct le *leu;
	struct le *lec;
	struct call *call = NULL;
	int err;

	for (leu = list_head(uag_list()); leu; leu = leu->next) {
		struct ua *ua = leu->data;
		for (lec = list_head(ua_calls(ua)); lec; lec = lec->next) {
			call = lec->data;

			if (call_state(call) == CALL_STATE_ESTABLISHED &&
					!call_is_onhold(call))
				goto out;
		}
	}

out:
	return call;
}
```

The second is shorter. And currently we have no other use-case for the call filter.

**Comment by cspiel1** [comment]
> List element has `void *data`. How about creating a new list with new elements and take the data to element from the ua calls list?

We could add a function that calls a callback function for each call:

```
typedef void (call_process_h)(struct call *call, void *arg);
int uag_process_calls(call_process_h, void *arg);
int uag_filter_calls(call_process_h, enum call_state st, void *arg);
```

But also here the benefit for this PR would be minimal. And what use cases do we have else?

**Comment by cspiel1** [comment]
Please look at the newest commit! If you prefer a solution with
```
int uag_filter_calls(call_process_h processh, enum call_state st, void *arg);
```
let me know!

**Comment by juha-h** [comment]
The more general one would be better if it is not possible to return the call list.

**Comment by cspiel1** [comment]
The two latest commits:
- The first adds the filter function uag_filter_calls() requested by @juha-h.
- The second adds a generic find function uag_find_call() similar to the filter which simplified menu by 13 lines.

Should we keep both functions?

**Comment by cspiel1** [comment]
Should we replace the uag_find_call_state() we added shortly also by the generic uag_find_call()? Here are the uses:
```
$ cppgrep uag_find_call_state
./modules/menu/static_menu.c:106:		call = uag_find_call_state(CALL_STATE_INCOMING);
./modules/menu/static_menu.c:179:		call = uag_find_call_state(CALL_STATE_INCOMING);
./modules/menu/menu.c:577:			call = uag_find_call_state(state[i]);
```

**Comment by juha-h** [comment]
Christian Spielberger writes:
> The two latest commits:

> - The first adds the filter function uag_filter_calls() requested by @juha-h.
> - The second adds a generic find function uag_find_call() similar to the filter which simplified menu by 13 lines.
> 
> Should we keep both functions?

The latter does not look like a good API call to me, since if there is
more than one call that matches the filter then it is not well defined,
which one is returned.

**Comment by cspiel1** [comment]
Same for uag_find_call_state(). So I will replace both by use of uag_filter_calls(). But please let me traverse the calls of one User-Agent in reverse order, like I changed in last commit!

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Same for uag_find_call_state(). So I will replace both by use of
> uag_filter_calls(). But please let me traverse the calls of one
> User-Agent in reverse order, like I changed in last commit!

I noticed the commit "ua: reverse order of the calls", which changed the
order in which calls are processed/looked for.

The order should not matter in case all matching calls are processed,
but if only one is returned like in `uag_find_call` then the right order
may depend on the application that is using the function.  And as I
mentioned earlier, I don't consider such an API call a good idea.

I'm still not convinced, why a new call list cannot be returned.  Call
objects themselves would not need to be copied, just pointers to them.

One solution to avoid this problem is just to have ua_calls() in the
API.  Apps could then call it on all relevant UAs and process the calls
of each.

**Comment by cspiel1** [comment]
In re/baresip a list of a struct X always is implemented by adding a list-element field to the struct.
```
struct X {
    struct le le;
    ...
};


```
Then an append looks like this:
```
struct X *x;
x = mem_zalloc(sizeof(x), NULL);
list_append(&list, &x->le, x);
```

A le can only be appended to one list. If you want a second independent list, you need another le in the struct.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> In re/baresip a list of a struct X always is implemented by adding a
> list-element field to the struct.

OK, got it.

Is call_id of call unique?  If yes, then the list could be list of
call_ids, since we already have these:
```
const char   *call_id(const struct call *call);
struct call  *call_find_id(const struct list *calls, const char *id);
```

**Comment by cspiel1** [comment]
Then we would need a new struct.
```
struct Idelem {
    struct le le;
    char callid;
}
```
And code in call.c or ua.c to build such lists. Belief me, the current variant with the callback function is much cleaner.

Have to add another commit that makes use of uag_filter_calls() again in menu.c. Ist not used currently. That was a mistake.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Then we would need a new struct.
> ```
> struct Idelem {
>     struct le le;
>     char callid;
> }
> ```

> And code in call.c or ua.c to build such lists. Belief me, the current
> variant with the callback function is much cleaner.

Callback based function is OK when something is done on all matching
elements.  But if more than one element matches and something is done on
only one arbitrary one, it is not a clean, generic solution.

> Have to add another commit that makes use of uag_filter_calls() again
> in menu.c. Ist not used currently. That was a mistake.

I'm already a bit lost with all the commits.  It would be easier to
digest if there would a separate pull request for API changes that would
not include any module related changes.

**Comment by cspiel1** [comment]
> I'm already a bit lost with all the commits.  It would be easier to
digest if there would a separate pull request for API changes that would
not include any module related changes.

You're right. I very cleanly did the first 8 commits. The rest where a lot of back-forward steps. I will do a rebase to make it easier to review this PR commit-by-commit.

**Comment by cspiel1** [comment]
This is the API change. https://github.com/baresip/baresip/pull/1293

**Comment by cspiel1** [comment]
This PR is replaced by: https://github.com/baresip/baresip/pull/1294

**Pull request closed without merging**